### DISCUSSION TITLE
OpenDT will nicht mit dem dritten MPPT.

### DISCUSSION THREAD

**Discussion by hoschiking** [hoylabs/OpenDTU-OnBattery#1344] [Q&A] [answered]
Nach langem testen habe ich den finalen Aufbau angefangen und möchte alle 3 MPPT per VE.Direct mit dem ESP verbinden.
2 gleichzeitig funktionieren immer, ein dritter will nicht.
Egal wie ich die Kabel und MPPT tausche, 2 funktionieren immer, der dritte (am RX3 Pin im Pin Mapping) will nicht.
Den RX3 an verschiedenen GPIO´s (5, 15, 18, 21 usw) getestet, nichts hilft.
Schliesse ich den MPPT vom RX3 an RX2, wird der sofort erkannt und ich kann die Daten sehen.
Alles durchgemssen, neue Kabel gebaut, nichts hilft.
Hat jemand eine Idee?
Habe das neueste Release, alle 3 MPPT sind per ADUM galvanisch getrennt.

**Comment by hoschiking** [comment]
Mittlerweile bin ich ein stück weiter gekommen.
Es scheint ein Fehler in OpenDTUoB zu sein, egal wich ich die Pins konfiguriere und welches Gerät ich anschliesse,
RX1 und RX2 gehen, RX3 nicht. Es sei denn, ich übersehe ewas.
Jemand eine Idee?

**Comment by schlimmchen** [comment, marked as answer]
Das klingt mir schwer danach als hättest du [die Doku](https://opendtu-onbattery.net/hardware/limitations/#hardware-uarts) nicht gelesen.

* Hast du einen ESP32-S3?
* Nutzt du `generic_esp32s3_usb` (das `_usb` am Ende ist entscheidend)?

Bitte schau nach, was dein Board an Nachrichten ausspuckt beim Starten. Der "SerialPortManager" sagt sehr deutlich, wer welchen Hardware-Serial-Port bekommt und wird vermutlich auch sagen, dass kein dritter für den dritten Victron zur Verfügung steht.

**Reply by hoschiking** [reply]
Oh ja, ich habe die Doku schon gelesen, aber nicht den Bereich "Limitations" 
Ich benutze zur Zeit einen einfachen ESP32 mit 4MB, somit "generic_esp32_4mb_no_ota"
Mein Fehler also.
Die ersten Versuche waren mit einem S3, habe dann auch "generic_esp32s3_usb" benutzt, damit hat es funktioniert.
Dieses ESP ist mittlerweile direkt am Speicher angeschlossen, da ich die Installtion aus diversen Gründen auf 2 ESP aufgteilt habe.
Da hat es natürlich funktioniert......
Man man man...... so schnell tappt man in eine Falle.
Heißt es dann auch, wenn ich zusätzlich noch ein JK-BMS anschliessen, kann ich auch an einem S3 nur max 2 MPPT anschliessen?
Ich dachte bis jetzt, dass für die weiteren VE.Direct SoftwareSerial verwendet wird, das hatte ich in irgend einem Beitrag mal gelesen.
Auch da lag ich falsch.
Danke dir für den schnellen Support

**Comment by schlimmchen** [comment]
> Heißt es dann auch, wenn ich zusätzlich noch ein JK-BMS anschliessen, kann ich auch an einem S3 nur max 2 MPPT anschliessen?

Leider korrekt.

> Ich dachte bis jetzt, dass für die weiteren VE.Direct SoftwareSerial verwendet wird

Das hatte ich evaluiert und befunden, dass das funktionieren würde, aber nur mit erheblichem Aufwand, weshalb ich das bisher nicht angegangen bin.

**Reply by hoschiking** [reply]
Mal eine andere Frage:
ein weiterer S3 ist bei Ali bestellt, dauert noch paar Tage.
Jetzt habe ich die Doku zur Verwendung von UART0 gelesen. Meiner Meinung nach spricht
prinzipiell nichts dagegen diesen Port für externe Hardware zu nutzen, solange diese robust genug ist.
Da ich den TX0 ohnehin für VE.Direct nicht benutze, beschränkt sich die Frage auf den RX.
Der ESP scheint es nicht zu stören beim Booten, mehrfach wiederholt, kein Ding.
Also habe ich den RX3 auf den Pin 3 (RX0 des ESP) gelegt. Funktioniert weiterhin nicht.
Ist das eine Besonderheit von OpenDTUoB, dass der RX0 generell nicht nutzbar ist?

**Reply by schlimmchen** [reply]
> Ist das eine Besonderheit von OpenDTUoB, dass der RX0 generell nicht nutzbar ist?

Weder ist der nicht generell nicht nutzbar, noch ist das OpenDTU-OnBattery-spezifisch.

Um welchen ESP32 geht es denn nun hier? ESP32-S3 mit `generic_esp32s3_usb`? Dann funktionieren auch GPIOs 43 und 44 "normal", aber eben erst nachdem der Bootloader fertig ist. Den Bootloader kann man auch dazu bringen, keine Ausgaben zu machen. Da gibt es einen Strapping-Pin. Steht im Datenblatt. Ist ein "advanced Feature", wenn man so will.

Weil du "Pin 3" erwähnt hast und damit vermutlich "GPIO 3" meinst, sprichst du wohl von einem ESP32 ohne -S3, und da ist UART0 immer in Verwendung. So steht es ja auch in der [Doku](https://opendtu-onbattery.net/hardware/limitations/#amount).

**Reply by hoschiking** [reply]
Ja ich meinte den ESP32 ohne S3, da der zur Zeit in Vernwedung ist.
Ja so steht es in der Doku, die du verlinkt hast, habs auch gelesen.
Ich habe mir aber auch eine andere Doku über den ESP32 angeschaut und generell ist UART0 auch für Peripherie verwendbar, nur eben erst nach dem Booten.  siehe:
[https://randomnerdtutorials.com/esp32-pinout-reference-gpios/](url)
Screenshot:
![2024-10-25 11_28_33-ESP32 Pinout Reference_ Which GPIO pins should you use_ _ Random Nerd Tutorials ](https://github.com/user-attachments/assets/3bcc8aa8-d1b7-48e3-970b-a4c2c3144bb4)
Aber du hast es jetzt ausreichend beantwortet.
Danke

**Reply by schlimmchen** [reply]
> generell ist UART0 auch für Peripherie verwendbar

Das ist natürlich richtig, und insofern hast du Recht, dass es OpenDTU-spezifisch ist. Das würde aber erfordern, dass wir eine Firmwarevariante anbieten ohne serielles Logging, und da bin ich absolut nicht begesitert von (Mehraufwand an Code für etwas, das ich nicht will, weil jeder seriell loggen könnten sollte, wenn Fehler gesucht werden).

**Reply by hoschiking** [reply]
Alles gut, darauf wollte ich nicht hinaus, ich wollte nur die Mechanismen dahinter verstehen, bei den Dingen bin ich immer noch ein fortgeschrittener Anfänger ;-)
Wie gesagt ein S3 ist bestellt, dann hat sich das Thema sowieso erledigt

**Comment by hoschiking** [comment]
OK, am Speicher ist der S3 und der wird nur fürs Funkmodul und BMS verwendet.
Ich muss also nur die beiden ESP tauschen, dann sollte es klappen.
Leider haben die unterschiedliche Anzahl von Pins, die Verdrahtung haut dann nicht mehr hin.
Ich muss also einen Tag nach dem Neuaufbau es wieder ändern.....
